### PR TITLE
proposed rewording of the Ord instance documentation

### PR DESCRIPTION
**Pull request by tmcl**
Re issue #250 

The Ord instance for HashMap depends on:

* toList', which flattens(?) the HashMap into a list of Leafs and Collisions
* The Semigroup instance for Ordering:
```
instance Semigroup Ordering where
    LT <> _ = LT
    EQ <> y = y
    GT <> _ = GT
```
Therefore, the ordering of two hashmaps depends on the ordering of the first non-equal pair.

Concerning the comparison:
* If the elements being compared are both Leafs, then the hashes are compared and if EQ, the keys are compared and if EQ, the values are compared.
* If the elements being compared are both Collisions, then the hashes are compared and if EQ, the key/value array is compared for length, and if EQ, then it uses unorderedCompare to compare the key/value pairs.
* A Leaf is less than a Collision, and a Collision is greater than a Leaf. 

I assume that the entry that will be the first entry will also depend on the hash.

Therefore, the ordering depends primarily on the hash. 

So it seems to me that the note intends to communicate:

> /Note:/ The ordering of elements in 'HashMap' depends primarily on the keys' hashes. Therefore, it inherits any of the properties of the hashes. In particular, if the hash in use is inconsistent between library versions, OSes or architectures, then the ordering will be inconsistent in the same circumstance.

I have included this proposed rewording in the documentation. I think it is good to call out that it depends primarily on the hash, since the instance is `(Ord k, Ord v) => Ord (HashMap k v)`. A person who reads this might think "oh, it is using the keys and values to determine the order", whereas in reality it is only using the keys and values to determine the order of two hashmaps whose early entries have the same hash but different keys or values. I suppose this isn't so unlikely, but it isn't as important as the conditions on the instance might make you think.

An alternative rewording was included in the original bug report. It is included here with modifications: 

> The actual order of elements in 'HashMap' and the result of
> `compare` may not b consistent because the hash is not guaranteed to be
> consistent across library versions, OSes and architectures

The main change here is that I have changed the word "stable" for "consistent", since I don't believe the author had in mind stability as it is usually understood in ordering (i.e. if you sort a sorted list of hashmaps, it is unchanged). The reference to library versions, OSes and architectures, as well as the hash, makes me believe the original author had it mind the possibility that automatically derived instances of Data.Hashable may change from version to version. (Is that true? It strikes me as a little surprising.)

I don't have a strong understanding of this library or `hashable`. I haven't actually used either - I came across this documentation while trying to understand another library.

### CODE REVIEW

File: Data/HashMap/Base.hs [hunk 1]
```diff
@@ -354,9 +354,11 @@ instance Ord k => Ord1 (HashMap k) where
...
 -- | The order is total.
 --
--- /Note:/ Because the hash is not guaranteed to be stable across library
--- versions, OSes, or architectures, neither is an actual order of elements in
--- 'HashMap' or an result of `compare`.is stable.
+-- /Note:/ The ordering of elements in 'HashMap' depends primarily on the keys'
+-- hashes. Therefore, it inherits any of the properties of the hashes. In
+-- particular, if the hash in use is inconsistent between library versions,
+-- OSes or architectures, then the ordering will be inconsistent in the same
+-- circumstance.
```
Comment (sjakobi) [inline Data/HashMap/Base.hs:361]:
```suggestion
-- circumstances.
```

(Non-native English speaker here. Sorry if this not an improvement!)

**Comment by tibbe** [comment]
The intent was to avoid users relying on the hash functions not changing in the future. Java made the mistake of specifying the exact hash function in the standard and has regretted it since. If users want stable hash functions (e.g. to compute file digests) they should not use hashable.

I forgot the version bump policy I use when changing the hash function.

File: Data/HashMap/Base.hs (hunk 1 above)
Comment (emilypi) [inline Data/HashMap/Base.hs:361]: Proposing a language change that makes it clear in both yours and @sjakobi's suggestion: 
 
```suggestion
-- OSes or architectures, then the ordering will reflect these inconsistencies. 
```

**Comment by sjakobi** [comment]
BTW, I've found some more discussion on the _stability_ of `Hashable` hashes in https://github.com/tibbe/hashable/issues/139.

**Comment by sjakobi** [comment]
Thanks, this wording sounds great to me!

@treeowl, could you give this a look too?

**Comment by treeowl** [comment]
Ehhhh.... I'm not a big fan of the old or new wording, because it gives more detail than is useful and pins the lack of specification to one feature when it's actually somewhat broader than that. I think we want to say something more like

> The ordering is total, relative to `(==)`. However, nothing else about the ordering is specified, and it may change from version to version of either this package or of `hashable`.

We may want to mention that this means, for example, that deserializing a `Set` of hashmaps using `fromDistinctAscList` is unsafe if the set was serialized using a different version of `unordered-containers` or of `hashable`.

**Comment by tmcl** [comment]
I think I like @treeowl 's proposal better. (Should I incorporate it into this PR?)

The additional comments probably deserve to go on `fromDistinctAscList` or somewhere that deals with overarching concerns, because the fact that `fromDistinctAscList` is unsafe across versions does not depend on comparing two HashMaps, but on comparing two hashes.

**Comment by sjakobi** [comment]
> > The ordering is total, relative to `(==)`.

I honestly don't know what that means! ;) Can this  be expressed in formulas like those for the [`Ord` docs](http://hackage.haskell.org/package/base-4.14.0.0/docs/Data-Ord.html#t:Ord)?

**Comment by treeowl** [comment]
> > > The ordering is total, relative to `(==)`.
> 
> I honestly don't know what that means! ;) Can this be expressed in formulas like those for the [`Ord` docs](http://hackage.haskell.org/package/base-4.14.0.0/docs/Data-Ord.html#t:Ord)?

This gets to the substitutivity issue. But maybe it's better, as I suggested elsewhere, to flip the substitutivity issue over to the other side.

**Comment by sjakobi** [comment]
Ah! Would something like

> total order, consistent with the Eq instance:
>
> compare x y == EQ = x == y

capture the same meaning then? That's something that I'd think more people would understand.

**Comment by sjakobi** [comment]
Here's a tweaked version of @treeowl's suggestion, that I hope would be slightly clearer:

> The ordering is total and consistent with the `Eq` instance. However, nothing else about the ordering is specified, and it may change from version to version of either this package or of hashable.

@emilypi, @tmcl, @treeowl Is this good? If not, could you possibly suggest another way to clarify the "relative to `(==)`" bit in the version from https://github.com/haskell-unordered-containers/unordered-containers/pull/259#issuecomment-638582857.

**Comment by treeowl** [comment]
Yeah, that sounds good.

**Comment by sjakobi** [comment]
@tmcl Could you update this PR with the wording above?

I think this is the last thing we should merge before the release (#266).

**Comment by tmcl** [comment]
Done.

**Comment by sjakobi** [comment]
Thank you @tmcl! :)

**Pull request merged**